elasticapm/utils/time.py

```diff
@@ -0,0 +1,81 @@
+#  BSD 3-Clause License
+#
+#  Copyright (c) 2019, Elasticsearch BV
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#
+#  * Redistributions of source code must retain the above copyright notice, this
+#    list of conditions and the following disclaimer.
+#
+#  * Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions and the following disclaimer in the documentation
+#    and/or other materials provided with the distribution.
+#
+#  * Neither the name of the copyright holder nor the names of its
+#    contributors may be used to endorse or promote products derived from
+#    this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+#  FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+#  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+#  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+#  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+import time
+
+CLOCK_DIFF = None
+
+
+def time_to_perf_counter(timestamp):
+    """
+    This function converts a given epoch timestamp in seconds (typically from
+    `time.time()`) to the "equivalent" result from `time.perf_counter()`.
+
+    Note that because these functions vary in their resolution and tick rate,
+    this is only a close approximation.
+
+    Note also that because `time.time()` is *usually* monotonic (but not
+    guaranteed), if a system clock is changed, this function could become
```

Comment:
Would it make sense to periodically update `CLOCK_DIFF` (e.g. on every nth call of `time_to_perf_counter`)?

Reply:
Maybe so. For lambda it's not necessary, because the runtime is not long-lived. But if we're going to use something like this for the otel PR that's open, we'll probably want to be able to invalidate previous findings.

Reply:
Decided to make it based on the time elapsed, rather than on the number of calls. I set it to 5 minutes.

Again, it should never fall out of sync unless someone changes the system time which I don't see happening on most modern systems. But it's worth handling that case since this is pretty lightweight. In this case only 5 minutes of data will ever be out of sync.